cmd/thanos/main.go

```diff
@@ -161,9 +170,81 @@ func joinCluster(logger log.Logger, typ cluster.PeerType, bindAddr, advertiseAdd
 	// TODO(fabxc): generate human-readable but random names?
 	name, err := ulid.New(ulid.Now(), rand.New(rand.NewSource(time.Now().UnixNano())))
 	if err != nil {
-		return nil, err
+		return cluster.JoinConfig{}, err
+	}
+	return cluster.JoinConfig{
+		Name:          name.String(),
+		Addr:          bindAddr,
+		AdvertiseAddr: advertiseAddr,
+		KnownPeers:    resolvedPeers,
+	}, nil
+}
+
+func resolvePeers(ctx context.Context, peers []string, myAddress string, res net.Resolver, waitIfEmpty bool) ([]string, error) {
+	var resolvedPeers []string
+
+	for _, peer := range peers {
+		host, port, err := net.SplitHostPort(peer)
+		if err != nil {
+			return nil, errors.Wrapf(err, "split host/port for peer %s", peer)
+		}
+
+		retryCtx, cancel := context.WithCancel(ctx)
+		ips, err := res.LookupIPAddr(ctx, host)
+		if err != nil {
+			// Assume direct address.
+			resolvedPeers = append(resolvedPeers, peer)
+			continue
+		}
+
+		if len(ips) == 0 {
+			var lookupErrSpotted bool
+
+			err := runutil.Retry(2*time.Second, retryCtx.Done(), func() error {
+				if lookupErrSpotted {
+					cancel()
+				}
```

Comment:
I suppose there's a reason for calling `cancel` here instead of just replacing `lookupErrSpotted = true` with it?
 Feels like a doc comment could help here.